gt/main.c

```diff
@@ -870,14 +794,344 @@ print_unsent_policy(struct ggu_policy *policy)
+	int num_to_send = 0;
+	int num_sent;
+	unsigned int i;
+
+	if (instance->num_ggu_pkts == 0)
+		return;
+
+	for (i = 0; i < max_pkts; i++) {
+		if (instance->ggu_pkts[i].buf == NULL)
+			continue;
+
+		prep_notify_pkt(&instance->ggu_pkts[i]);
+		bufs[num_to_send++] = instance->ggu_pkts[i].buf;
+	}
+
+	num_sent = rte_eth_tx_burst(gt_conf->net->front.id,
+		instance->tx_queue, bufs, num_to_send);
+	if (unlikely(num_sent < num_to_send)) {
+		/*
+		 * Need to exclude the first @num_sent valid entries
+		 * found in instance->ggu_pkts from being printed.
+		 */
+		int sent_count = 0;
+		for (i = 0; i < max_pkts; i++) {
+			if (instance->ggu_pkts[i].buf == NULL)
+				continue;
+
+			if (sent_count < num_sent) {
+				sent_count++;
+				continue;
+			}
+
+			print_unsent_policies(&instance->ggu_pkts[i]);
+			rte_pktmbuf_free(instance->ggu_pkts[i].buf);
+		}
+	}
+
+	for (i = 0; i < max_pkts; i++) {
+		if (instance->ggu_pkts[i].buf == NULL)
+			continue;
+
+		instance->ggu_pkts[i].buf = NULL;
+		instance->num_ggu_pkts--;
+	}
+
+	RTE_VERIFY(instance->num_ggu_pkts == 0);
+}
+
+/*
+ * Start building a new notification packet for the Gatekeeper
+ * server indicated by @pkt_info.
+ *
+ * If there's no more room for a notification packet, then
+ * send a random one to make room.
+ */
+static struct ggu_notify_pkt *
+add_notify_pkt(struct gt_config *gt_conf, struct gt_instance *instance,
+	struct gt_packet_headers *pkt_info)
+{
+	unsigned int max_pkts = gt_conf->max_ggu_notify_pkts;
+	struct ggu_notify_pkt *ggu_pkt = NULL;
+	unsigned int lcore_id = rte_lcore_id();
+	unsigned int i;
+
+	/* Find an available packet, sending a packet if necessary. */
+	if (instance->num_ggu_pkts == max_pkts) {
+		int idx = rte_rand() % max_pkts;
+		for (i = 0; i < max_pkts; i++) {
+			if (instance->ggu_pkts[idx].buf != NULL)
+				break;
+			idx = (idx + 1) % max_pkts;
+		}
+		RTE_VERIFY(i < max_pkts);
+
+		ggu_pkt = &instance->ggu_pkts[idx];
+		send_notify_pkt(gt_conf, instance, ggu_pkt);
+	} else {
+		for (i = 0; i < max_pkts; i++) {
+			if (instance->ggu_pkts[i].buf == NULL) {
+				ggu_pkt = &instance->ggu_pkts[i];
+				break;
+			}
+		}
+	}
+	RTE_VERIFY(ggu_pkt != NULL);
+
+	ggu_pkt->ipaddr.proto = pkt_info->outer_ethertype;
+	if (ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv4) {
+		struct ipv4_hdr *ipv4_hdr =
+			(struct ipv4_hdr *)pkt_info->outer_l3_hdr;
+		ggu_pkt->ipaddr.ip.v4.s_addr = ipv4_hdr->src_addr;
+	} else if (likely(ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv6)) {
+		struct ipv6_hdr *ipv6_hdr =
+			(struct ipv6_hdr *)pkt_info->outer_l3_hdr;
+		rte_memcpy(ggu_pkt->ipaddr.ip.v6.s6_addr, ipv6_hdr->src_addr,
+			sizeof(ggu_pkt->ipaddr.ip.v6.s6_addr));
+	} else {
+		rte_panic("Unexpected condition: gt at lcore %u adding to notification packet to Gatekeeper server with unknown IP version %hu\n",
+			lcore_id, ggu_pkt->ipaddr.proto);
+	}
+
+	ggu_pkt->buf = rte_pktmbuf_alloc(
+		gt_conf->net->gatekeeper_pktmbuf_pool[
+			rte_lcore_to_socket_id(lcore_id)]);
```

Comment:
I think we can drop variable `lcore_id` and `rte_lcore_to_socket_id()` in favor of `rte_socket_id()`.